### PR TITLE
chore(edit-page-v2): Add scroll while dragging content inside editor

### LINKED ISSUE #28469: Edit Page 2: Allow Users to scroll while dragging
### Parent Issue

https://github.com/dotCMS/core/issues/26644

### Task

- [ ] Enable the user to scroll inside the iframe page while dragging from palette, outside of browser like dot-asset or move contentlets 

### Proposed Objective

Same as Parent Issue

### Proposed Priority

Priority 2 - Important

### Acceptance Criteria

_No response_

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.

_No response_

### Assumptions & Initiation Needs

_No response_

### Quality Assurance Notes & Workarounds

_No response_

### Sub-Tasks & Estimates

_No response_

### PR DESCRIPTION
**Pull request by KevinDavilaDotCMS**
### Proposed Changes
* Enable scroll in UVE when dragging element from palette, dot-asset out browser or moving inside editor between containers

In VTL:

https://github.com/dotCMS/core/assets/144152756/76efdcce-076a-48ce-b361-a5b57f2f84bd

In Headless:

https://github.com/dotCMS/core/assets/144152756/6d10c111-df2b-4e8f-bbde-90c81df0aacc

### CODE REVIEW

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html [hunk 1]
```diff
@@ -71,12 +71,13 @@
...
-                    !es.editorData.device &&
-                    es.state === editorState.DRAGGING
+                    (es.editorData.canEditVariant &&
+                        !!es.bounds.length &&
+                        !es.editorData.device &&
+                        es.state === editorState.DRAGGING) ||
+                    es.state === editorState.SCROLLING
                 "
-                [containers]="es.bounds"
+                [containers]="es.state === editorState.SCROLLING ? [] : es.bounds"
```
Comment (zJaaal) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html:80]: if you can go from `SCROLLING` to `DRAGGING` you don't need this logic

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts [hunk 2]
```diff
@@ -406,6 +420,42 @@ export class EditEmaEditorComponent implements OnInit, OnDestroy {
...
+                if (!direction) {
+                    this.store.updateEditorState(EDITOR_STATE.DRAGGING);
+
+                    return;
+                }
+
+                this.store.updateEditorState(EDITOR_STATE.SCROLLING);
+
+                this.iframe.nativeElement.contentWindow?.postMessage(
+                    { name: 'scroll-inside-iframe', direction },
```
Comment (zJaaal) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:456]: create an entry in the `NOTIFY_CUSTOMER` an rename it to `ema-scroll-page` or something like that

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts [hunk 3]
```diff
@@ -304,6 +314,15 @@ export class EditEmaEditorComponent implements OnInit, OnDestroy {
...
 
     handleDragEvents() {
+        this.store.currentState$.pipe(takeUntil(this.destroy$)).subscribe((state) => {
+            if (state === EDITOR_STATE.DRAGGING) {
+                this.iframe.nativeElement.contentWindow?.postMessage(
+                    NOTIFY_CUSTOMER.EMA_REQUEST_BOUNDS,
+                    this.host
+                );
+            }
+        });
```
Comment (zJaaal) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:324]: Why did you move it from the dragStart?

**Review by zJaaal — CHANGES_REQUESTED** [review]

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts (hunk 3 above)
Comment (KevinDavilaDotCMS) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:324]: For perfomance
Before, the request bound was sent in the dragstart

Now with the scroll, during the dragover we can send the request bound or the request scroll

And if we sent the request bound during the dragover, many calls were made and it could affect the performance of the browser

That's why in the dragover only the state changes and the request bound is sent once

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html (hunk 1 above)
Comment (KevinDavilaDotCMS) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html]: U right, added that logic to updater!

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts (hunk 2 above)
Comment (KevinDavilaDotCMS) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts]: Added!

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html [hunk 4]
```diff
@@ -71,10 +71,11 @@
...
                 *ngIf="
-                    es.editorData.canEditVariant &&
-                    !!es.bounds.length &&
-                    !es.editorData.device &&
-                    es.state === editorState.DRAGGING
+                    (es.editorData.canEditVariant &&
+                        !!es.bounds.length &&
+                        !es.editorData.device &&
+                        es.state === editorState.DRAGGING) ||
+                    es.state === editorState.SCROLLING
```
Comment (zJaaal) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html:78]: This `or`  is misplaced and it looks like you don't need it

**Review by zJaaal — CHANGES_REQUESTED** [review]

File: core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts [hunk 5]
```diff
@@ -677,6 +675,19 @@ export class EditEmaStore extends ComponentStore<EditEmaState> {
...
+     *
+     * @memberof EditEmaStore
+     */
+    readonly setScrollingState = this.updater((state) => {
+        return {
+            ...state,
+            editorState: EDITOR_STATE.SCROLLING,
+            bounds: []
+        };
+    });
```
Comment (fmontes) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts:689]: `setScrollingState` vs `updateEditorScrollState` weird, not sure if is the naming.

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html (hunk 4 above)
Comment (fmontes) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html:78]: These conditionals are just out crazy by now, really difficult to maintain or even understand.

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts [hunk 6]
```diff
@@ -304,6 +314,15 @@ export class EditEmaEditorComponent implements OnInit, OnDestroy {
     }
 
     handleDragEvents() {
+        this.store.currentState$.pipe(takeUntil(this.destroy$)).subscribe((state) => {
+            if (state === EDITOR_STATE.DRAGGING) {
```
Comment (fmontes) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:318]: Don't we have a scrolling selector?

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts [hunk 7]
```diff
@@ -406,6 +420,42 @@ export class EditEmaEditorComponent implements OnInit, OnDestroy {
...
             .subscribe((event: DragEvent) => {
                 event.preventDefault(); // Prevent file opening
+
+                const iframeRect = this.iframe.nativeElement.getBoundingClientRect();
+
+                const isInsideIframe =
+                    event.clientX > iframeRect.left && event.clientX < iframeRect.right;
+
+                let direction;
+
```
Comment (fmontes) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:430]:
```
if (!isInsideIframe) {
    return
}
```

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts [hunk 8]
```diff
@@ -499,19 +549,28 @@ export class EditEmaEditorComponent implements OnInit, OnDestroy {
...
+                if (this.isVTLPage()) {
                     // If is VTL, the content is updated by store.code$
-                    this.reloadIframe();
+                    this.setIframeContent(curr.code);
                 } else {
-                    this.setIframeContent(code);
+                    // Only reload if is Headless.
+                    this.reloadIframe();
                 }
             });
```
Comment (fmontes) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:575]: Should we have another selector?

File: dotCMS/src/main/webapp/html/js/editor-js/sdk-editor.esm.js
Comment (fmontes) [inline dotCMS/src/main/webapp/html/js/editor-js/sdk-editor.esm.js:1]: Off topic, but we need to minimize and uglify this.

File: core-web/libs/sdk/client/src/lib/editor/listeners/listeners.ts [hunk 9]
```diff
@@ -80,6 +80,11 @@ export function listenEditorMessages() {
                 break;
             }
         }
+
+        if (event.data.name === NOTIFY_CUSTOMER.EMA_SCROLL_INSIDE_IFRAME) {
+            const scrollY = event.data.direction === 'up' ? -120 : 120;
```
Comment (fmontes) [inline core-web/libs/sdk/client/src/lib/editor/listeners/listeners.ts:85]: We might be able to release with this BUT this number can be more or less based on how close of the edge the drag event is.

**Review by fmontes — CHANGES_REQUESTED** [review]

**Review by zJaaal — CHANGES_REQUESTED** [review]

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html (hunk 4 above)
Comment (oidacra) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html:78]: Pretty hard to read. Is there a way to use a computed signal with a good name?

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts [hunk 10]
```diff
@@ -304,6 +314,15 @@ export class EditEmaEditorComponent implements OnInit, OnDestroy {
     }
 
     handleDragEvents() {
+        this.store.currentState$.pipe(takeUntil(this.destroy$)).subscribe((state) => {
```
Comment (oidacra) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:317]: instead `this.destroy$ `use `takeUntilDestroyed()`

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts [hunk 11]
```diff
@@ -499,19 +549,28 @@ export class EditEmaEditorComponent implements OnInit, OnDestroy {
     handleReloadContent() {
         this.store.contentState$
             .pipe(
+                startWith({ state: EDITOR_STATE.LOADING, code: '' }),
```
Comment (oidacra) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:552]: instead `this.destroy$ `use `takeUntilDestroyed()`

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html (hunk 4 above)
Comment (zJaaal) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html:78]: Maybe these conditions should be selectors in the store. There are a couple of these that are repeated

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts (hunk 6 above)
Comment (KevinDavilaDotCMS) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:318]: No, Scroll is a EDITOR_STATE item
Maybe we can create a new selector with filter, but I don't think it's necessary now

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html (hunk 4 above)
Comment (KevinDavilaDotCMS) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html]: Moved conditionals to store!

File: core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts [hunk 12]
```diff
@@ -259,7 +257,27 @@ export class EditEmaStore extends ComponentStore<EditEmaState> {
...
+                    !editorData.device &&
+                    editor.page.canEdit &&
+                    (currentState === EDITOR_STATE.IDLE ||
+                        currentState === EDITOR_STATE.DRAGGING) &&
+                    !editorData.page.isLocked,
+                showDropzone:
+                    editorData.canEditVariant &&
+                    !editorData.device &&
+                    (currentState === EDITOR_STATE.DRAGGING ||
+                        currentState === EDITOR_STATE.SCROLLING),
```
Comment (zJaaal) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts:273]: do we want to show the dropzone while scrolling?

Comment (zJaaal) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts:273]: Should we check the bounds here?

**Review by zJaaal — CHANGES_REQUESTED** [review]

File: core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts (hunk 12 above)
Comment (KevinDavilaDotCMS) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts:273]: Yes, we need to have the dropzone
If the dropzone is hidden, the dragover is lost
The dropzone is rendered when scrolling but with empty bounds array

For that, is not needed check bounds, because with the states we already know that (Dragging = normal bounds, scrolling = empty array bounds)

**Review by zJaaal — APPROVED** [review]

**Review by fmontes — APPROVED** [review]

**Review by oidacra — APPROVED** [review]
